### ISSUE TITLE
Misleading example about "Postconditions" on page about context functions

### ISSUE THREAD

**Issue by jrudolph** [lampepfl/dotty#12332]
https://dotty.epfl.ch/docs/reference/contextual/context-functions.html#example-postconditions says:

> Since WrappedResult is an opaque type alias, its values need not be boxed, and since ensuring is added as an extension method, its argument does not need boxing either. Hence, the implementation of ensuring is as about as efficient as the best possible code one could write by hand:
> ```scala
> val s =
>   val result = List(1, 2, 3).sum
>   assert(result == 6)
>   result
> ```

"Is about as efficient" doesn't seem to be true since it still allocates and executes a closure. The only extra feature it adds is offering the `ensuring(result == ...)` syntax. You can get exactly the same cost using `def ensuring(condition: T => Boolean): T = { assert(condition(x)); x }` and calling it `.ensuring(_ == 6)`.

If you want less cost you should use `inline` (which works both for the regular syntax and the `result == ` one if you remove the opaque type).

**Comment by OlivierBlanvillain** [comment]
I agree, here is what CFR outputs for the two `val s = ...` mentioned in the text:

```scala
// val s = List(1, 2, 3).sum.ensuring(result == 6)
s = BoxesRunTime.unboxToInt((Object)Foo.PostConditions.ensuring(
    ((IterableOnceOps).List()
      .apply((Seq)ScalaRunTime.wrapIntArray(new int[]{1, 2, 3})))
      .sum((Numeric)Numeric.IntIsIntegral$.MODULE$),
     (JFunction1.mcZI.sp & Serializable)evidence$1 ->
      BoxesRunTime.unboxToInt((Object)Foo.PostConditions
        .result(BoxesRunTime.boxToInteger((int)evidence$1))) == 6));
```

```scala
// val s =
//   List(1, 2, 3).sum
//   assert(result == 6)
//   result
s = BoxesRunTime.unboxToInt((Object)((IterableOnceOps).List()
      .apply((Seq)ScalaRunTime.wrapIntArray(new int[]{1, 2, 3})))
      .sum((Numeric)Numeric.IntIsIntegral$.MODULE$));
if (result != 6) {
    throw Scala3RunTime.assertFailed();
}
```
